### ISSUE TITLE
请问是否方便上传基于websocket通信的代码

### ISSUE THREAD

**Issue by hasegetc** [kuafuwang/LspCpp#6]
* 是否方便上传基于websocket通信的代码
* 或者简单描述下使用的websocket框架或者库？
* 我需要使用websocket的通信方式，但对C++不甚了解，求指导，谢谢

**Comment by kuafuwang** [comment]
请等等.还在写.

**Comment by hasegetc** [comment]
* 很期待，楼主辛苦了！可以弄个大概，我也许也能出点力

**Comment by kuafuwang** [comment]
> * 很期待，楼主辛苦了！可以弄个大概，我也许也能出点力



> * 是否方便上传基于websocket通信的代码
> * 或者简单描述下使用的websocket框架或者库？
> * 我需要使用websocket的通信方式，但对C++不甚了解，求指导，谢谢
https://github.com/kuafuwang/LspCpp/blob/master/WebsocketExample.cpp
这是一个例子, 我用的是boost库. 他里面有websocket的库.
也不知道对不对,因为我没有和别的语言写的测试过,就是自己写了server和client来测试.

**Comment by kuafuwang** [comment]
https://github.com/kuafuwang/LspCpp/issues/1  你以前问过类似的问题吧. 我刚才翻了一下,你为什么说java lsp里面有websocket的.你说的是 https://github.com/eclipse/lsp4j 这个吗? 这个应该没有websocket的. vscode用的肯定也不是websocket, 这个用普通的socket 也行,为什么要websocket呢?

**Comment by hasegetc** [comment]
> #1 你以前问过类似的问题吧. 我刚才翻了一下,你为什么说java lsp里面有websocket的.你说的是 https://github.com/eclipse/lsp4j 这个吗? 这个应该没有websocket的. vscode用的肯定也不是websocket, 这个用普通的socket 也行,为什么要websocket呢?

* 楼主你好，谢谢你的作品，很漂亮。我瞪着这份源码两天，也没想明白怎么在websocket中实现，在这里像您好好学习：）
* 我们有份C++的编译器，一个小众语言的。需要做成云平台。之前是做一些预研，现在开始动手，由于web前端对websocket支持较好。所以选中了websocket + jsonrpc作为基本通信方式
* 感觉在云领域，还是websocket用的多一点，比如：https://github.com/theia-ide/typescript-language-server、https://github.com/TypeFox/vscode-ws-jsonrpc
* https://github.com/eclipse/lsp4j 中有如下包： org.eclipse.lsp4j.websocket.jakarta，应该是websocket的
* 再次感谢，像您致敬

**Comment by hasegetc** [comment]
> > * 很期待，楼主辛苦了！可以弄个大概，我也许也能出点力
> 
> > * 是否方便上传基于websocket通信的代码
> > * 或者简单描述下使用的websocket框架或者库？
> > * 我需要使用websocket的通信方式，但对C++不甚了解，求指导，谢谢
> >   https://github.com/kuafuwang/LspCpp/blob/master/WebsocketExample.cpp
> >   这是一个例子, 我用的是boost库. 他里面有websocket的库.
> >   也不知道对不对,因为我没有和别的语言写的测试过,就是自己写了server和client来测试.

楼主你好，你的代码直接运行是可以的。不过，我直接运行代码中的server端，然后通过如下工具：http://coolaf.com/tool/chattest ， 连接到：ws://127.0.0.1:9333/ ，能完成连接，但发送数据时，server无法接收到，在WebSocketServer::do_accept和RemoteEndPoint::consumer中，断点均无法触发，想请教是不是服务端完成websocket连接后，很快又关闭了？

**Comment by kuafuwang** [comment]
![16](https://user-images.githubusercontent.com/3636678/112596481-d78b0800-8e46-11eb-8884-741b060970e7.png)
可以通信的.不过这个是只接受 [lsp 格式的  JSON-RPC ](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#headerPart) 才会触发 RemoteEndPoint::consumer 的.那个聊天工具因为没有办法发送回车,所有没有办法发送一个完整的Language Server Protocol 格式的消息的.

**Comment by kuafuwang** [comment]
> WebSocketServer::do_accept

WebSocketServer::do_accept 在连接的时候触发.

**Comment by kuafuwang** [comment]
[boost 的websocket的例子](https://www.boost.org/doc/libs/1_75_0/libs/beast/example/websocket/)
我猜测你是想要一个c++版本的websocket通信的例子,接收到你指定的  JSON-RPC  格式的后,反序列化它吧.你并不需要lsp库.你只是需要websocket库和JSON-RPC 序列化的库.

**Comment by kuafuwang** [comment]
![17](https://user-images.githubusercontent.com/3636678/112600360-11124200-8e4c-11eb-96ec-94f44093cba8.png)
你看看这个效果.估计你想要的是这个.

**Comment by kuafuwang** [comment]
`
void do_session(tcp::socket socket)
{
    try
    {
     
        websocket::stream<tcp::socket> ws{ std::move(socket) };

        // Set a decorator to change the Server of the handshake
        ws.set_option(websocket::stream_base::decorator(
            [](websocket::response_type& res)
            {
                res.set(http::field::server,
                    std::string(BOOST_BEAST_VERSION_STRING) +
                    " websocket-server-sync");
            }));

        // Accept the websocket handshake
        ws.accept();

        for (;;)
        {
            // This buffer will hold the incoming message
            beast::flat_buffer buffer;
        	

            // Read a message
            ws.read(buffer);
            char* data = reinterpret_cast<char*>(buffer.data().data());
            std::string elements(data, data + buffer.data().size());
            std::cout << elements << std::endl;
            lsp::Any any;
            any.SetJsonString(elements, rapidjson::Type::kObjectType);
            td_initialize::response rsp;
            try
            {
                any.Get(rsp);
                std::cout << "反序列成功" << std::endl;
            }
            catch (...)
            {
                std::cout << "反序列失败" << std::endl;
            }
           
            // Echo the message back
            ws.text(ws.got_text());
            ws.write(buffer.data());
        }
    }
    catch (beast::system_error const& se)
    {
        // This indicates that the session was closed
        if (se.code() != websocket::error::closed)
            std::cerr << "Error: " << se.code().message() << std::endl;
    }
    catch (std::exception const& e)
    {
        std::cerr << "Error: " << e.what() << std::endl;
    }
}

//------------------------------------------------------------------------------

int main(int argc, char* argv[])
{

   
    try
    {
       
        auto const address = net::ip::make_address("127.0.0.1");
        auto const port = static_cast<unsigned short>(std::atoi("9333"));

        // The io_context is required for all I/O
        net::io_context ioc{ 1 };

        // The acceptor receives incoming connections
        tcp::acceptor acceptor{ ioc, {address, port} };
        for (;;)
        {
            // This will receive the new connection
            tcp::socket socket{ ioc };

            // Block until we get a connection
            acceptor.accept(socket);

            // Launch the session, transferring ownership of the socket
            std::thread(
                &do_session,
                std::move(socket)).detach();
        }
    }
    catch (const std::exception& e)
    {
        std::cerr << "Error: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }
}

`

**Comment by hasegetc** [comment]
是的，谢谢楼主，我好好学习一下，获益匪浅，万分感谢：）
